Update Go to 1.15

tklauser opened the pull request:
This also reduces binary size quite significantly (mostly due to the new linker):

Go 1.15:
```
-rwxrwxr-x 1 tklauser tklauser 7.0M Aug 12 12:38 bugtool/cilium-bugtool
-rwxrwxr-x 1 tklauser tklauser  46M Aug 12 12:38 cilium/cilium
-rwxrwxr-x 1 tklauser tklauser  14M Aug 12 12:38 cilium-health/cilium-health
-rwxrwxr-x 1 tklauser tklauser  55M Aug 12 12:38 daemon/cilium-agent
-rwxrwxr-x 1 tklauser tklauser  28M Aug 12 12:38 hubble-relay/hubble-relay
-rwxrwxr-x 1 tklauser tklauser  55M Aug 12 12:38 operator/cilium-operator
-rwxrwxr-x 1 tklauser tklauser  53M Aug 12 12:38 operator/cilium-operator-aws
-rwxrwxr-x 1 tklauser tklauser  46M Aug 12 12:38 operator/cilium-operator-azure
-rwxrwxr-x 1 tklauser tklauser  44M Aug 12 12:38 operator/cilium-operator-generic
-rwxrwxr-x 1 tklauser tklauser  12M Aug 12 12:38 plugins/cilium-cni/cilium-cni
-rwxrwxr-x 1 tklauser tklauser  15M Aug 12 12:38 plugins/cilium-docker/cilium-docker
```

Go 1.14.7:

```
-rwxrwxr-x 1 tklauser tklauser 8.6M Aug 12 13:03 bugtool/cilium-bugtool
-rwxrwxr-x 1 tklauser tklauser  49M Aug 12 13:03 cilium/cilium
-rwxrwxr-x 1 tklauser tklauser  15M Aug 12 13:03 cilium-health/cilium-health
-rwxrwxr-x 1 tklauser tklauser  58M Aug 12 13:03 daemon/cilium-agent
-rwxrwxr-x 1 tklauser tklauser  31M Aug 12 13:04 hubble-relay/hubble-relay
-rwxrwxr-x 1 tklauser tklauser  60M Aug 12 13:03 operator/cilium-operator
-rwxrwxr-x 1 tklauser tklauser  58M Aug 12 13:03 operator/cilium-operator-aws
-rwxrwxr-x 1 tklauser tklauser  49M Aug 12 13:04 operator/cilium-operator-azure
-rwxrwxr-x 1 tklauser tklauser  47M Aug 12 13:03 operator/cilium-operator-generic
-rwxrwxr-x 1 tklauser tklauser  14M Aug 12 13:04 plugins/cilium-cni/cilium-cni
-rwxrwxr-x 1 tklauser tklauser  17M Aug 12 13:04 plugins/cilium-docker/cilium-docker
```

rolinh commented on cilium-docker-plugin.Dockerfile:
```diff
@@ -1,6 +1,6 @@
 # (first line comment needed for DOCKER_BUILDKIT use)
 #
-FROM docker.io/library/golang:1.14.7 as builder
+FROM docker.io/library/golang:1.15 as builder
```
```suggestion
FROM docker.io/library/golang:1.15.0 as builder
```
I think that for the Go Docker image, the `.0` needs to be appended. If not. when Go 1.15.1 is released it'll be automatically picked up which is probably not what we want.

tklauser replied:
Good point, thanks. Since this was updated automatically based on `GO_VERSION` using `make update-golang` I'll have to adjust these manually. We want to keep `1.15` where we don't refer to `docker.io/library/golang`.

aanm reviewed (changes requested):
```patch
diff --git a/go.mod b/go.mod
index 52b7ac350..c2adecd2f 100644
--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module github.com/cilium/cilium
 
-go 1.14
+go 1.15
 
 // direct dependencies
 require (
```

We also need to update the dev documentation (I think we list golang there with 1.14)

edit: Also the dockerfiles in `images/` need to be updated.

tklauser committed:
images: add script to update all Go toolchain images and run it

Copied from github.com/cilium/image-tools and adjusted to directory
structure in github.com/cilium/cilium

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
Update Go to 1.15

Use docker.io/library/golang:1.15.0 for the Go Docker image to avoid
automatically picking up new 1.15.x releases.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
images: set DEBIAN_FRONTEND=noninteractive

Avoid any apt configuration prompts.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

kaworu commented on images/scripts/update-golang-image.sh:
```diff
@@ -0,0 +1,33 @@
+used_by=($(git grep -l GOLANG_IMAGE= images/{builder,runtime}))
+
+for i in "${used_by[@]}" ; do
+  sed "s|\(GOLANG_IMAGE=\)docker.io/library/golang:.*\$|\1${image}@${image_digest}|" "${i}" > "${i}.sedtmp" && mv "${i}.sedtmp" "${i}"
```
nit: use `sed -i` here instead of the sedtmp dance.

tklauser replied:
Good point. Since this script was copied from https://github.com/cilium/image-tools I'd address this there first and then vendor it in a follow-up PR.

tklauser replied:
See cilium/image-tools#58
